workers/unity/Packages/io.improbable.gdk.core/Systems/EntitySystem.cs

```diff
@@ -1,61 +1,127 @@
-using System.Collections.Generic;
+using System;
+using Unity.Collections;
 using Unity.Entities;
-using Unity.Profiling;
+using Unity.Jobs;
 
 namespace Improbable.Gdk.Core
 {
     [DisableAutoCreation]
     [AlwaysUpdateSystem]
     [UpdateInGroup(typeof(SpatialOSReceiveGroup.InternalSpatialOSReceiveGroup))]
-    [UpdateBefore(typeof(SpatialOSReceiveSystem))]
-    public class EntitySystem : ComponentSystem
+    [UpdateAfter(typeof(SpatialOSReceiveSystem))]
+    public class EntitySystem : SystemBase
     {
-        public int ViewVersion { get; private set; }
-
-        private readonly List<EntityId> entitiesAdded = new List<EntityId>();
-        private readonly List<EntityId> entitiesRemoved = new List<EntityId>();
-
-        private ProfilerMarker applyDiffMarker = new ProfilerMarker("EntitySystem.ApplyDiff");
-
-        public List<EntityId> GetEntitiesAdded()
+        private struct EntityCollection : IDisposable
         {
-            return entitiesAdded;
-        }
+            public NativeArray<EntityId> EntityIds { get; private set; }
+            public int EntityCount { get; private set; }
+            public JobHandle JobHandle { get; set; }
 
-        public List<EntityId> GetEntitiesRemoved()
-        {
-            return entitiesRemoved;
-        }
-
-        internal void ApplyDiff(ViewDiff diff)
-        {
-            using (applyDiffMarker.Auto())
+            public EntityCollection(int size)
             {
-                entitiesAdded.Clear();
-                entitiesRemoved.Clear();
+                EntityIds = new NativeArray<EntityId>(size, Allocator.Persistent);
+                EntityCount = 0;
+                JobHandle = default;
+            }
 
-                // todo decide on a container and remove this
-                foreach (var entityId in diff.GetEntitiesAdded())
+            public NativeSlice<EntityId> Slice
+            {
+                get
                 {
-                    entitiesAdded.Add(entityId);
+                    JobHandle.Complete();
+                    return new NativeSlice<EntityId>(EntityIds, 0, EntityCount);
                 }
+            }
 
-                foreach (var entityId in diff.GetEntitiesRemoved())
+            public void EnsureSize(ref EntityQuery query)
+            {
+                EntityCount = query.CalculateEntityCount();
+                if (EntityIds.Length >= EntityCount)
                 {
-                    entitiesRemoved.Add(entityId);
+                    return;
                 }
 
-                if (entitiesAdded.Count != 0 || entitiesRemoved.Count != 0)
+                EntityIds.Dispose();
+                EntityIds = new NativeArray<EntityId>(EntityCount << 1, Allocator.Persistent);
```

Comment:
Since we do resizes, would a `NativeList` not work better? It can use the NativeList `GetEnumerator()` instead of the NativeSlice to make iteration nicely without write access.